htlcswitch/link.go

```diff
@@ -2292,228 +2292,20 @@ func (l *channelLink) processRemoteAdds(fwdPkg *channeldb.FwdPkg,
-			// within our local state update log, then send the
-			// update entry to the remote party.
-			//
-			// NOTE: We make an exception when the value requested
-			// by the invoice is zero. This means the invoice
-			// allows the payee to specify the amount of satoshis
-			// they wish to send.  So since we expect the htlc to
-			// have a different amount, we should not fail.
-			if !l.cfg.DebugHTLC && invoice.Terms.Value > 0 &&
-				pd.Amount < invoice.Terms.Value {
-
-				log.Errorf("rejecting htlc due to incorrect "+
-					"amount: expected %v, received %v",
-					invoice.Terms.Value, pd.Amount)
-
-				failure := lnwire.NewFailUnknownPaymentHash(
-					pd.Amount,
-				)
-				l.sendHTLCError(
-					pd.HtlcIndex, failure, obfuscator, pd.SourceRef,
-				)
-
-				needUpdate = true
-				continue
-			}
-
-			// As we're the exit hop, we'll double check the
-			// hop-payload included in the HTLC to ensure that it
-			// was crafted correctly by the sender and matches the
-			// HTLC we were extended.
-			//
-			// NOTE: We make an exception when the value requested
-			// by the invoice is zero. This means the invoice
-			// allows the payee to specify the amount of satoshis
-			// they wish to send.  So since we expect the htlc to
-			// have a different amount, we should not fail.
-			if !l.cfg.DebugHTLC && invoice.Terms.Value > 0 &&
-				fwdInfo.AmountToForward < invoice.Terms.Value {
-
-				log.Errorf("Onion payload of incoming htlc(%x) "+
-					"has incorrect value: expected %v, "+
-					"got %v", pd.RHash, invoice.Terms.Value,
-					fwdInfo.AmountToForward)
-
-				failure := lnwire.NewFailUnknownPaymentHash(
-					pd.Amount,
-				)
-				l.sendHTLCError(
-					pd.HtlcIndex, failure, obfuscator, pd.SourceRef,
-				)
-
-				needUpdate = true
-				continue
-			}
-
-			// We'll also ensure that our time-lock value has been
-			// computed correctly.
-			expectedHeight := heightNow + minCltvDelta
-			switch {
-			case !l.cfg.DebugHTLC && pd.Timeout < expectedHeight:
-				log.Errorf("Incoming htlc(%x) has an "+
-					"expiration that is too soon: "+
-					"expected at least %v, got %v",
-					pd.RHash[:], expectedHeight, pd.Timeout)
-
-				failure := lnwire.FailFinalExpiryTooSoon{}
-				l.sendHTLCError(
-					pd.HtlcIndex, failure, obfuscator,
-					pd.SourceRef,
-				)
-
-				needUpdate = true
-				continue
-
-			case !l.cfg.DebugHTLC && pd.Timeout != fwdInfo.OutgoingCTLV:
-				log.Errorf("HTLC(%x) has incorrect "+
-					"time-lock: expected %v, got %v",
-					pd.RHash[:], pd.Timeout,
-					fwdInfo.OutgoingCTLV)
-
-				failure := lnwire.NewFinalIncorrectCltvExpiry(
-					fwdInfo.OutgoingCTLV,
-				)
-				l.sendHTLCError(
-					pd.HtlcIndex, failure, obfuscator, pd.SourceRef,
-				)
-
-				needUpdate = true
-				continue
-			}
-
-			preimage := invoice.Terms.PaymentPreimage
-			err = l.channel.SettleHTLC(
-				preimage, pd.HtlcIndex, pd.SourceRef, nil, nil,
+			updated, err := l.processExitHop(
```

Comment:
Non blocker: let's also do the forwarding hop as well while we're at it?

Reply:
It is a sensible idea, but I don't want to extend the PR further now.